pkg/ensign/events.go

```diff
@@ -17,9 +23,63 @@ func (s *Server) Publish(stream api.Ensign_PublishServer) (err error) {
 	o11y.OnlinePublishers.Inc()
 	defer o11y.OnlinePublishers.Dec()
 
+	// Parse the context for authentication information
+	ctx := stream.Context()
+	claims, ok := contexts.ClaimsFrom(ctx)
+	if !ok {
+		// NOTE: this should never happen because the interceptor will catch it, but
+		// this check prevents nil panics and guards against regressions.
+		sentry.Fatal(ctx).Msg("no claims available in publish stream")
+		return status.Error(codes.Unauthenticated, "missing credentials")
+	}
+
+	// Verify that the user has permissions to publish.
+	if !claims.HasPermission(permissions.Publisher) {
+		log.Warn().Msg("attempt to open publisher stream without publisher permission")
+		return status.Error(codes.Unauthenticated, "not authorized to perform this action")
+	}
+
+	var projectID ulid.ULID
+	if projectID, err = ulids.Parse(claims.ProjectID); err != nil || ulids.IsZero(projectID) {
+		sentry.Warn(ctx).Err(err).Msg("could not parse projectID from claims")
+		return status.Error(codes.Unauthenticated, "not authorized to perform this action")
+	}
+
+	// Get the topic IDs this user is allowed to publish to
+	// TODO: make this more efficient and globalize to prevent inconsistencies
+	allowedTopics := make(map[string]struct{})
+	topics := s.meta.ListTopics(projectID)
+
+	for topics.Next() {
+		// Get and parse the key for the topic ID
+		var key meta.ObjectKey
+		copy(key[:], topics.Key())
+
+		// Parse the ULID:
+		var topicID ulid.ULID
+		if topicID, err = key.ObjectID(); err != nil {
+			sentry.Error(ctx).Err(err).Msg("could not parse topic id")
+			topics.Release()
+			return status.Error(codes.Internal, "could not open publisher stream")
+		}
+
+		allowedTopics[topicID.String()] = struct{}{}
+	}
+
+	if err := topics.Error(); err != nil {
+		sentry.Warn(ctx).Err(err).Msg("could not fetch topics")
+		topics.Release()
+		return status.Error(codes.Internal, "could not open publisher stream")
+	}
+	topics.Release()
```

Comment:
Is there a reason why the release needs to be done immediately rather than putting it on the defer stack?

Reply:
Oh I see there's a wg.wait() later on.

Reply:
I've moved this functionality to a shared method to ensure that our cursors are cleaned up correctly.